botocore/awsrequest.py

```diff
@@ -534,6 +557,84 @@ def reset_stream(self):
             raise UnseekableStreamError(stream_object=self.body)
 
 
+class AWSRequestCompressor:
+    """A class that can compress the body of an ``AWSRequest``."""
+
+    def compress(self, body, encodings, headers):
+        """Compresses the reqeust body using the specified encodings.
+
+        Set or append the `Content-Encoding` header with the matched encoding
+        if not present.
+        """
+        for encoding in encodings:
+            encoder = SUPPORTED_COMPRESSION_ENCODINGS.get(encoding)
+            if encoder is not None:
+                if 'Content-Encoding' not in headers:
+                    headers['Content-Encoding'] = encoding
+                elif encoding not in headers['Content-Encoding']:
+                    headers.replace_header(
+                        'Content-Encoding',
+                        f'{headers["Content-Encoding"]},{encoding}',
+                    )
+                return self._compress_body(body, encoder)
+            else:
+                logger.debug('Unsupported compression encoding: %s' % encoding)
+        return body
+
+    def _compress_body(self, body, encoder):
+        if isinstance(body, str):
+            return encoder(body.encode('utf-8'))
+        elif isinstance(body, (bytes, bytearray)):
+            return encoder(body)
+        elif hasattr(body, 'read'):
```

Comment:
What's the motivation for this specific check? How does it compare to, for example, `isinstance(io.IOBase)` ([IOBase docs](https://docs.python.org/3/library/io.html#io.IOBase))? (Not asking for a change, just trying to understand the reasoning.)